### PR TITLE
#324 implement centralized memory constraint

### LINKED ISSUE #324: Implement centralized memory-constraint algorithm from C++ code

### PR DESCRIPTION
**Pull request by lifflander**
Fixes #324 

Still very much a draft. Need to figure out better way for comparison operator.

### CODE REVIEW

File: src/lbaf/Execution/lbsCentralizedPrefixOptimizerAlgorithm.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+                tally_assigned += 1
+            else:
+                if not(len(min_rank.get_shared_blocks()) >= self._max_shared_ids and not sid in min_rank.get_shared_blocks()):
+                    # more efficient way to do this?
+                    min_ranks.append(min_rank)
+                    heapq.heapify(min_ranks)
+
+                tally_rejected += 1
+
+        self._logger.info("try bin: {}, size={}, id={}; assigned={}, rejected={}\n".format(tbin, size, sid, tally_assigned, tally_rejected))
```
Comment (ppebay) [inline src/lbaf/Execution/lbsCentralizedPrefixOptimizerAlgorithm.py:203]: In LBAF we try to use f-strings everywhere for code style consistency

**Review by ppebay — COMMENTED** [review]
First pass with suggestions as this PR is still `WIP`

File: src/lbaf/Applications/LBAF_app.py [hunk 2]
```diff
@@ -375,14 +379,14 @@ def main(self):
...
 
-        # Instantiate phase to VT file writer if started from a log file
-        if "data_stem" in self.params.__dict__:
-            vt_writer = VTDataWriter(
-                curr_phase,
-                self.logger,
-                self.params.output_file_stem,
-                output_dir=self.params.output_dir)
-            vt_writer.write()
+        # # Instantiate phase to VT file writer if started from a log file
```
Comment (lifflander) [inline src/lbaf/Applications/LBAF_app.py:382]: This needs to be fixed

File: src/lbaf/Utils/logger.py [hunk 3]
```diff
@@ -34,17 +34,17 @@ def logger(name: str = "root", conf: str = None):
         config = {}
 
     # Assign formatting properties
-    clr_fnc = black if config.get("terminal_background") == "light" else light_white
+    clr_fnc = yellow if config.get("terminal_background") == "light" else light_white
```
Comment (lifflander) [inline src/lbaf/Utils/logger.py:37]: I need to fix this also

File: src/lbaf/Applications/LBAF_app.py (hunk 2 above)
Comment (ppebay) [reply src/lbaf/Applications/LBAF_app.py:382]: Yes and done in PR #346

File: src/lbaf/Utils/logger.py (hunk 3 above)
Comment (ppebay) [reply src/lbaf/Utils/logger.py:37]: Agree, created #347 for that purpose

**Review by ppebay — APPROVED** [review]
PR approved but creating follow-on issue for documentation #348

**Comment by ppebay** [comment]
Reviewed/approved -- @lifflander please feel free to merge if you do not intend to make additional changes to this PR.

File: src/lbaf/Applications/LBAF_app.py [hunk 4]
```diff
@@ -371,6 +371,10 @@ def run(self):
             self.__logger.info("No brute force optimization performed")
             a_min_max = []
 
+        # Perform central prefix optimizer when requested
+        if self.__parameters.algorithm["name"] != "CentralizedPrefixOptimizer":
```
Comment (ppebay) [inline src/lbaf/Applications/LBAF_app.py:375]: I don't understand why logging `Starting centralized prefix optimizer` if the algorithm is **not** the `CentralizedPrefixOptimizer`

File: src/lbaf/Model/lbsRank.py [hunk 5]
```diff
@@ -59,6 +59,9 @@ def copy(self, rank):
         self.__sentinel_objects = copy.copy(rank.__sentinel_objects)
         self.__migratable_objects = copy.copy(rank.__migratable_objects)
 
+    def __lt__(self, other):
```
Comment (ppebay) [inline src/lbaf/Model/lbsRank.py:62]: I am not sure that we want this to be always the comparison operator between ranks, as they can be sorted by other QoIs (like index, work, etc.)

**Review by ppebay — CHANGES_REQUESTED** [review]
Thanks for the change; I made some other comments however

File: src/lbaf/Execution/lbsCentralizedPrefixOptimizerAlgorithm.py [hunk 6]
```diff
@@ -0,0 +1,309 @@
+import sys
+import math
+import itertools
+import heapq
+from logging import Logger
+
+from .lbsAlgorithmBase import AlgorithmBase
+from ..Model.lbsPhase import Phase
```
[the remark this answers is not in the document]
Comment (ppebay) [reply src/lbaf/Execution/lbsCentralizedPrefixOptimizerAlgorithm.py:7]: Resolved

File: src/lbaf/Execution/lbsCentralizedPrefixOptimizerAlgorithm.py [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+
+        # Prepare input data for rank order enumerator
+        self._logger.info(f"Starting optimizer")
+
+        phase_ranks = self._phase.get_ranks()
+
+        max_shared_ids = 0
+        for rank in phase_ranks:
+            max_shared_ids = max(len(rank.get_shared_block_ids()), max_shared_ids)
+        self._max_shared_ids = max_shared_ids + 1
```
[the remark this answers is not in the document]
Comment (ppebay) [reply src/lbaf/Execution/lbsCentralizedPrefixOptimizerAlgorithm.py:48]: Reorganized constructor to make this clearer

File: src/lbaf/Applications/LBAF_app.py (hunk 4 above)
Comment (ppebay) [reply src/lbaf/Applications/LBAF_app.py]: Removed this logging line which in any event indicates the opposite of what it means

File: src/lbaf/Model/lbsRank.py (hunk 5 above)
Comment (ppebay) [reply src/lbaf/Model/lbsRank.py:62]: Leaving this as default comparison but creating follow-on issue for later discussion of how we want to handle this in general (and be consistent with vt in this regard)

**Review by ppebay — APPROVED** [review]
Made requested changes

**Comment by lifflander** [comment]
Your changes look good to me. I can't approve it because it's my PR.

Were you going to to change the `<` operator to be more general than just compare load?

**Comment by ppebay** [comment]
@lifflander #412

**Pull request merged**